File: agenttel-node/src/anomaly/patterns.ts
import { IncidentPattern } from '../enums-impl.js';
import type { RollingSnapshot } from '../interfaces.js';

interface LatencyTrend {
  samples: number[];
  times: number[];
}

/** Detects known incident patterns from telemetry observations. */
export class PatternMatcher {
  private readonly latencyDegradationThreshold: number;
  private readonly errorRateSpikeThreshold: number;
  private readonly cascadeFailureMinServices: number;
  private depErrors = new Map<string, number>();
  private latencyTrends = new Map<string, LatencyTrend>();

  constructor(latencyDegradation = 2.0, errorRateSpike = 0.1, cascadeMin = 3) {
    this.latencyDegradationThreshold = latencyDegradation > 0 ? latencyDegradation : 2.0;
    this.errorRateSpikeThreshold = errorRateSpike > 0 ? errorRateSpike : 0.1;
    this.cascadeFailureMinServices = cascadeMin > 0 ? cascadeMin : 3;
  }

  recordDependencyError(dependency: string): void {
    this.depErrors.set(dependency, (this.depErrors.get(dependency) ?? 0) + 1);
  }

  recordLatency(operationName: string, latencyMs: number): void {
    let trend = this.latencyTrends.get(operationName);
    if (!trend) {
      trend = { samples: [], times: [] };
      this.latencyTrends.set(operationName, trend);
    }
    trend.samples.push(latencyMs);
    trend.times.push(Date.now());
    if (trend.samples.length > 100) {
      trend.samples = trend.samples.slice(-100);
      trend.times = trend.times.slice(-100);
    }
  }

  resetDependencyErrors(): void {
    this.depErrors.clear();
  }

  detectPatterns(
    operationName: string,
    currentLatencyMs: number,
    isError: boolean,
    snapshot: RollingSnapshot,
  ): IncidentPattern[] {
    const patterns: IncidentPattern[] = [];

    // CASCADE_FAILURE: multiple dependencies failing
    let failingDeps = 0;
    for (const count of this.depErrors.values()) {
      if (count > 0) failingDeps++;
    }
    if (failingDeps >= this.cascadeFailureMinServices) {
      patterns.push(IncidentPattern.CASCADE_FAILURE);
    }

    if (snapshot.sampleCount === 0) return patterns;

    // ERROR_RATE_SPIKE
    if (snapshot.errorRate > this.errorRateSpikeThreshold) {
      patterns.push(IncidentPattern.ERROR_RATE_SPIKE);
    }

    // LATENCY_DEGRADATION
    if (snapshot.mean > 0 && currentLatencyMs > snapshot.mean * this.latencyDegradationThreshold) {
      patterns.push(IncidentPattern.LATENCY_DEGRADATION);
    }

    // MEMORY_LEAK: monotonically increasing latency with rising errors
    const trend = this.latencyTrends.get(operationName);
    if (trend && trend.samples.length >= 10) {
      const recent = trend.samples.slice(-10);
      if (isMonotonicallyIncreasing(recent) && snapshot.errorRate > 0.01) {
        patterns.push(IncidentPattern.MEMORY_LEAK);
      }
    }

    return patterns;
  }
}

function isMonotonicallyIncreasing(vals: number[]): boolean {
  if (vals.length < 3) return false;
  let increases = 0;
  for (let i = 1; i < vals.length; i++) {
    if (vals[i] > vals[i - 1]) increases++;
  }
  return increases / (vals.length - 1) >= 0.8;
}

File: agenttel-go/anomaly/patterns.go
package anomaly

import (
	"sync"
	"time"

	"go.agenttel.dev/agenttel/enums"
	"go.agenttel.dev/agenttel/models"
)

// PatternMatcher detects known incident patterns from telemetry observations.
type PatternMatcher struct {
	mu sync.Mutex

	latencyDegradationThreshold float64
	errorRateSpikeThreshold     float64
	cascadeFailureMinServices   int

	// dependency error tracking
	depErrors map[string]int

	// latency trend tracking per operation
	latencyTrends map[string]*latencyTrend
}

type latencyTrend struct {
	samples []float64
	times   []time.Time
}

// NewPatternMatcher creates a pattern matcher with configurable thresholds.
func NewPatternMatcher(latencyDegradation, errorRateSpike float64, cascadeMin int) *PatternMatcher {
	if latencyDegradation <= 0 {
		latencyDegradation = 2.0
	}
	if errorRateSpike <= 0 {
		errorRateSpike = 0.1
	}
	if cascadeMin <= 0 {
		cascadeMin = 3
	}
	return &PatternMatcher{
		latencyDegradationThreshold: latencyDegradation,
		errorRateSpikeThreshold:     errorRateSpike,
		cascadeFailureMinServices:   cascadeMin,
		depErrors:                   make(map[string]int),
		latencyTrends:               make(map[string]*latencyTrend),
	}
}

// RecordDependencyError records an error for a dependency.
func (p *PatternMatcher) RecordDependencyError(dependency string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.depErrors[dependency]++
}

// RecordLatency records a latency observation for trend analysis.
func (p *PatternMatcher) RecordLatency(operationName string, latencyMs float64) {
	p.mu.Lock()
	defer p.mu.Unlock()
	t, ok := p.latencyTrends[operationName]
	if !ok {
		t = &latencyTrend{}
		p.latencyTrends[operationName] = t
	}
	t.samples = append(t.samples, latencyMs)
	t.times = append(t.times, time.Now())
	// Keep bounded
	if len(t.samples) > 100 {
		t.samples = t.samples[len(t.samples)-100:]
		t.times = t.times[len(t.times)-100:]
	}
}

// ResetDependencyErrors clears dependency error counters.
func (p *PatternMatcher) ResetDependencyErrors() {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.depErrors = make(map[string]int)
}

// DetectPatterns analyzes current telemetry to identify incident patterns.
func (p *PatternMatcher) DetectPatterns(
	operationName string,
	currentLatencyMs float64,
	isError bool,
	snapshot models.RollingSnapshot,
) []enums.IncidentPattern {
	p.mu.Lock()
	defer p.mu.Unlock()

	var patterns []enums.IncidentPattern

	// CASCADE_FAILURE: multiple dependencies failing
	failingDeps := 0
	for _, count := range p.depErrors {
		if count > 0 {
			failingDeps++
		}
	}
	if failingDeps >= p.cascadeFailureMinServices {
		patterns = append(patterns, enums.IncidentPatternCascadeFailure)
	}

	if snapshot.IsEmpty() {
		return patterns
	}

	// ERROR_RATE_SPIKE: sudden increase beyond threshold
	if snapshot.ErrorRate > p.errorRateSpikeThreshold {
		patterns = append(patterns, enums.IncidentPatternErrorRateSpike)
	}

	// LATENCY_DEGRADATION: sustained latency increase
	if snapshot.Mean > 0 && currentLatencyMs > snapshot.Mean*p.latencyDegradationThreshold {
		patterns = append(patterns, enums.IncidentPatternLatencyDegradation)
	}

	// MEMORY_LEAK: monotonically increasing latency with rising errors
	if t, ok := p.latencyTrends[operationName]; ok && len(t.samples) >= 10 {
		if isMonotonicallyIncreasing(t.samples[len(t.samples)-10:]) && snapshot.ErrorRate > 0.01 {
			patterns = append(patterns, enums.IncidentPatternMemoryLeak)
		}
	}

	return patterns
}

func isMonotonicallyIncreasing(vals []float64) bool {
	if len(vals) < 3 {
		return false
	}
	increases := 0
	for i := 1; i < len(vals); i++ {
		if vals[i] > vals[i-1] {
			increases++
		}
	}
	return float64(increases)/float64(len(vals)-1) >= 0.8
}
